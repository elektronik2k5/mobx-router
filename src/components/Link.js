import React from 'react';
import { observer, } from 'mobx-react';

const newTabProps = {
  target: '_blank',
  rel: 'noopener',
}

export function Link({
  view,
  className,
  params = {},
  queryParams = {},
  href = view.replaceUrlParams(params, queryParams),
  store = {},
  refresh = false,
  style,
  children,
  title = children,
  router = store.router,
  shouldOpenInNewTab = false,
  ...linkProps,
}){
  if (!router) {
    return console.error(
      "The 'router' prop must be defined for a Link component to work!", { args: arguments, }
    )
  }
  const anchorProps = {
    ...(style && { style, }),
    ...(className && { className, }),
    ...(shouldOpenInNewTab && newTabProps),
    onClick(e){
      const isMiddleClick = e.which === 2
      const isCmdOrCtrl = (e.metaKey || e.ctrlKey);
      const realShouldOpenInNewTab = shouldOpenInNewTab || isMiddleClick || isCmdOrCtrl;
      const shouldNavigateManually = refresh || realShouldOpenInNewTab || isCmdOrCtrl;
      if (!shouldNavigateManually) {
        e.preventDefault()
        router.goTo(view, params, store, queryParams)
      }
    },
    href,
    children: title,
    ...linkProps,
  }
  return <a {...anchorProps} />
}

export const ObserverLink = observer(Link)

export default ObserverLink
