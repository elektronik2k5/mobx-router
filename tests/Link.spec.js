import React from 'react'
import Route from '../src/route'
import Link from '../src/components/Link'
import renderer from 'react-test-renderer'

const view = new Route({
  path: '/profile/:username/:tab',
  component: <div/>,
})

function renderLinkToJson(props){
  return renderer.create(<Link {...props} />).toJSON()
}

test('Link renders', () => {
  const router = jest.fn()
  const tree = renderLinkToJson({ view, title: 'link', router, })
  expect(tree).toMatchSnapshot()
})

test('Link in new tab renders', () => {
  const router = jest.fn()
  const tree = renderLinkToJson({ view, title: 'link in new tab', router, shouldOpenInNewTab: true, })
  expect(tree).toMatchSnapshot()
})
