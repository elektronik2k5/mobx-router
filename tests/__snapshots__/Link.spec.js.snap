// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link in new tab renders 1`] = `
<a
  href="/profile"
  onClick={[Function]}
  rel="noopener"
  target="_blank"
>
  link in new tab
</a>
`;

exports[`Link renders 1`] = `
<a
  href="/profile"
  onClick={[Function]}
>
  link
</a>
`;
